### DISCUSSION TITLE
HOW TO: Create a Custom ActiveX Control with twinBASIC

### DISCUSSION THREAD

**Discussion by WaynePhillipsEA** [twinbasic/twinbasic#808] [Show and tell]
In case you missed it, twinBASIC can now create ActiveX controls!

See this great tutorial by @mwolf02 for a walkthrough...
## [HOW TO: Create a Custom ActiveX Control with twinBASIC](https://nolongerset.com/create-activex-control-with-twinbasic/)

<img src='https://user-images.githubusercontent.com/25738109/164957412-ecf9d94f-b570-474b-b782-9d0f43c4d9ba.jpg' width=300/>

> The Microsoft Access community is about to get a massive infusion of innovation thanks to the continued development of twinBASIC.
> 
> In the past, developing ActiveX controls–custom controls that you could insert onto a Microsoft Access form or report–required a completely different skill set than most Access developers had.  You could build them in VB6, but that meant they would never be able to work with 64-bit Access.  The other option was to build them in a different language entirely.  While certainly not impossible, it required learning new syntax and added friction to the process.
> 
> twinBASIC is about to change all of that.

**Comment by mwolfe02** [comment]
Funny story.  I was putting together my twinBASIC presentation for Access DevCon Vienna and under the heading "Future Possibilities" I added a bullet point for "Custom ActiveX Controls."  

About an hour later, I fire up the latest version of the IDE and right there on the splash screen I see something that says "ActiveX Control."  I thought, "Could it really be...?"  Fifteen minutes later I'm inserting a custom ActiveX control on an Access form.  I was blown away.

I felt bad writing an article about it because it seemed like overkill.  The whole process was so easy that I wasn't sure what value I could possibly add.  If nothing else, the article should bring some visibility to this amazing feature.

This could be an absolute game-changer.

**Reply by WaynePhillipsEA** [reply]
To be honest, I only brought the ActiveX control support work forward after you mentioned that you were presenting at the DevCon again this year.  I was concerned that you mightn't have enough new Access-related stuff to talk about (with regards to tB), so that got me thinking what I could do to help.

I wasn't 100% sure I could pull it off in time, but it looks like I did :)

Anyway, I'm glad you're liking it... and all tutorials are most welcome, as documentation is so sparse at the moment.  Thanks

**Reply by mwolfe02** [reply]
I was really pleased with how the demo portion worked out in my DevCon presentation (even with the slight hiccup with my dev environment when presenting live).  Hopefully, I got more than a few people's wheels turning.

**Reply by bclothier** [reply]
I totally agree. Wayne did a phenomenal job. 

I wanted to showcase the capabilities of twinBASIC a bit more, so I started a new project, [TwinBasicSevenZip](https://github.com/bclothier/TwinBasicSevenZip/) to do just that. Right now it's just methods but I hope to also provide UI to report progress on extracting/compressing 7-zip files.

**Reply by wqweto** [reply]
@bclothier I never overcame multi-threading issues with VB6 runtime in [VszLib](https://github.com/wqweto/VszLib) effort as some of the 7-zip formats performs I/O on separate threads (e.g. .rar) and these worker threads are **not** initialized for COM even (needed CoInitialize on callback entry).

Implementing such 7-zip wrapper will test TB runtime under extreme conditions IMO. It's very nice how declaring the interfaces with exact IIDs does not require a separate typelib and already your effort looks great success and I'm tempted to port VszLib to TB now but we'll see :-))

**Reply by bclothier** [reply]
I missed that in my original research. Good to know. I think tB should be able to handle it but we'll definitely put it to test. From my other projects, I know there is no issue using win32's `CreateThread` where VB6 would require additional hacks or workaround to make it sort of work. We'll get to find how far we can go!
